src/sunadaptcontroller/imexgus/sunadaptcontroller_imexgus.c

```diff
@@ -0,0 +1,222 @@
+  C->ops->seterrorbias   = SUNAdaptController_SetErrorBias_ImExGus;
+  C->ops->updateh        = SUNAdaptController_UpdateH_ImExGus;
+  C->ops->space          = SUNAdaptController_Space_ImExGus;
+
+  /* Create content */
+  content = NULL;
+  content = (SUNAdaptControllerContent_ImExGus)malloc(sizeof *content);
+  if (content == NULL)
+  {
+    (void) SUNAdaptController_Destroy(C);
+    return (NULL);
+  }
+
+  /* Attach content */
+  C->content = content;
+
+  /* Fill content with default/reset values */
+  SUNAdaptController_SetDefaults_ImExGus(C);
+  SUNAdaptController_Reset_ImExGus(C);
+
+  return (C);
+}
+
+/* -----------------------------------------------------------------
+ * Function to set ImExGus parameters
+ */
+
+int SUNAdaptController_SetParams_ImExGus(SUNAdaptController C,
+                                         sunrealtype k1e, sunrealtype k2e,
+                                         sunrealtype k1i, sunrealtype k2i)
+{
+  SACIMEXGUS_K1E(C) = k1e;
+  SACIMEXGUS_K2E(C) = k2e;
+  SACIMEXGUS_K1I(C) = k1i;
+  SACIMEXGUS_K2I(C) = k2i;
+  return SUNADAPTCONTROLLER_SUCCESS;
+}
+
+
+
+/* -----------------------------------------------------------------
+ * implementation of controller operations
+ * ----------------------------------------------------------------- */
+
+SUNAdaptController_Type SUNAdaptController_GetType_ImExGus(SUNAdaptController C)
+{ return SUN_ADAPTCONTROLLER_H; }
+
+int SUNAdaptController_EstimateStep_ImExGus(SUNAdaptController C, sunrealtype h,
+                                            int p, sunrealtype dsm, sunrealtype* hnew)
+{
+  /* order parameter to use */
+  const int ord = p + 1;
+
+  /* modified method for first step */
+  if (SACIMEXGUS_FIRSTSTEP(C))
+  {
+    /* set usable time-step adaptivity parameters */
+    const sunrealtype k = -RCONST(1.0) / ord;
+    const sunrealtype e = SUNMAX(SACIMEXGUS_BIAS(C) * dsm, TINY);
+
+    /* compute estimated optimal time step size */
+    *hnew = h * SUNRpowerR(e,k);
+  }
+  else
+  {
+    /* set usable time-step adaptivity parameters */
+    const sunrealtype k1e = -SACIMEXGUS_K1E(C) / ord;
+    const sunrealtype k2e = -SACIMEXGUS_K2E(C) / ord;
+    const sunrealtype k1i = -SACIMEXGUS_K1I(C) / ord;
+    const sunrealtype k2i = -SACIMEXGUS_K2I(C) / ord;
+    const sunrealtype e1 = SUNMAX(SACIMEXGUS_BIAS(C) * dsm, TINY);
+    const sunrealtype e2 = e1 / SUNMAX(SACIMEXGUS_EP(C), TINY);
+    const sunrealtype hrat = h / SACIMEXGUS_HP(C);
+
+    /* compute estimated optimal time step size */
+    *hnew = h * SUNMIN(hrat * SUNRpowerR(e1,k1i) * SUNRpowerR(e2,k2i),
+                       SUNRpowerR(e1,k1e) * SUNRpowerR(e2,k2e));
+  }
+
+  /* return with success */
+  return SUNADAPTCONTROLLER_SUCCESS;
+}
+
+int SUNAdaptController_Reset_ImExGus(SUNAdaptController C)
+{
+  SACIMEXGUS_EP(C) = RCONST(1.0);
+  SACIMEXGUS_FIRSTSTEP(C) = SUNTRUE;
+  return SUNADAPTCONTROLLER_SUCCESS;
+}
+
+int SUNAdaptController_SetDefaults_ImExGus(SUNAdaptController C)
+{
```

Comment:
```suggestion
{
  if (C == NULL) { return SUNADAPTCONTROLLER_ILL_INPUT; }
```

Reply:
Although it's not the typical use case, this function can be called directly so I think we should still check the inputs.